include/deal.II/trilinos/nox.templates.h

```diff
@@ -0,0 +1,1027 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2022 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_trilinos_nox_templates
+#define dealii_trilinos_nox_templates
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/trilinos/nox.h>
+
+#ifdef DEAL_II_WITH_TRILINOS
+
+#  include <NOX_Abstract_Group.H>
+#  include <NOX_Abstract_Vector.H>
+#  include <NOX_Solver_Factory.H>
+#  include <NOX_Solver_Generic.H>
+#  include <NOX_StatusTest_Combo.H>
+#  include <NOX_StatusTest_MaxIters.H>
+#  include <NOX_StatusTest_NormF.H>
+#  include <NOX_StatusTest_RelativeNormF.H>
+
+DEAL_II_NAMESPACE_OPEN
+
+#  ifndef DOXYGEN
+
+namespace TrilinosWrappers
+{
+  namespace internal
+  {
+    namespace NOXWrappers
+    {
+      template <typename VectorType>
+      class Group;
+
+      /**
+       * Implementation of the abstract interface
+       * NOX::Abstract::Vector for deal.II vectors. For details,
+       * see
+       * https://docs.trilinos.org/dev/packages/nox/doc/html/classNOX_1_1Abstract_1_1Vector.html.
+       */
+      template <typename VectorType>
+      class Vector : public NOX::Abstract::Vector
+      {
+      public:
+        /**
+         * Create empty vector.
+         */
+        Vector() = default;
+
+        /**
+         * Wrap an existing vector. The ownership is not transferred.
+         */
+        Vector(VectorType &vector)
+        {
+          this->vector.reset(&vector, [](auto *) { /*nothing to do*/ });
+        }
+
+        /**
+         * Initialize every element of this vector with gamma.
+         */
+        NOX::Abstract::Vector &
+        init(double gamma) override
+        {
+          *vector = gamma;
```

Comment:
Its a pity that `double` has to be used in this interface, rather than `VectorType::value_type`. Do you think that we should cast `gamma` to a `VectorType::value_type`, in case the underlying value is a `float` or `complex<float>`?

Ditto for `scale()`, `update()` etc.